File: src/agent/architect.ts
import { readOnlyToolFilter } from "../permissions/checker.ts";
import type { Provider } from "../provider/types.ts";
import type { ToolRegistry } from "../tools/registry.ts";
import type { Message } from "../types.ts";
import type { AgentLoopOptions } from "./loop.ts";
import { runAgentLoop } from "./loop.ts";
import type { AgentEvent } from "./types.ts";

export interface ArchitectOptions {
	onPlanReady?: (plan: string) => Promise<boolean>;
}

/**
 * Two-phase architect/editor pipeline.
 *
 * Phase 1 (architect): Runs with planMode and read-only tools to produce a plan.
 * Phase 2 (editor): Runs with the full tool set to execute the plan.
 *
 * Yields all events from both phases.
 */
export async function* runArchitectPipeline(
	architectProvider: Provider,
	editorProvider: Provider,
	registry: ToolRegistry,
	messages: Message[],
	options?: AgentLoopOptions & ArchitectOptions,
): AsyncGenerator<AgentEvent> {
	// Phase 1: Architect — plan mode with read-only tools
	const architectMessages: Message[] = [...messages];
	let plan: string | undefined;

	try {
		const architectGen = runAgentLoop(architectProvider, registry, architectMessages, {
			...options,
			planMode: true,
			toolFilter: readOnlyToolFilter,
		});

		for await (const event of architectGen) {
			yield event;
			if (event.type === "plan_complete") {
				plan = event.plan;
			}
		}
	} catch (err) {
		yield {
			type: "error",
			error: err instanceof Error ? err : new Error(String(err)),
		};
		return;
	}

	if (!plan) {
		yield {
			type: "error",
			error: new Error("Architect phase produced no plan"),
		};
		return;
	}

	// Check with onPlanReady callback
	if (options?.onPlanReady) {
		const approved = await options.onPlanReady(plan);
		if (!approved) {
			yield {
				type: "error",
				error: new Error("Plan was rejected"),
			};
			return;
		}
	}

	// Phase 2: Editor — full tool access, original messages + plan
	const editorMessages: Message[] = [
		...messages,
		{
			role: "user",
			content: `Execute the following plan:\n\n${plan}`,
		},
	];

	const editorGen = runAgentLoop(editorProvider, registry, editorMessages, {
		...options,
		planMode: false,
		toolFilter: undefined,
	});

	for await (const event of editorGen) {
		yield event;
	}
}

File: src/agent/architect.rs
//! Two-phase architect/editor pipeline.

use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

use async_stream::stream;
use futures::Stream;
use futures::StreamExt;

use super::loop_::{run_agent_loop, AgentLoopOptions};
use super::types::AgentEvent;
use crate::permissions::checker::read_only_tool_filter;
use crate::provider::types::Provider;
use crate::tools::registry::ToolRegistry;
use crate::types::{Message, UserMessage};

pub type OnPlanReady =
    Arc<dyn Fn(String) -> Pin<Box<dyn Future<Output = bool> + Send>> + Send + Sync>;

#[derive(Default, Clone)]
pub struct ArchitectOptions {
    pub on_plan_ready: Option<OnPlanReady>,
}

pub fn run_architect_pipeline<'a>(
    architect_provider: Arc<dyn Provider>,
    editor_provider: Arc<dyn Provider>,
    registry: ToolRegistry,
    messages: &'a mut Vec<Message>,
    loop_options: AgentLoopOptions,
    architect_options: ArchitectOptions,
) -> Pin<Box<dyn Stream<Item = AgentEvent> + Send + 'a>> {
    Box::pin(stream! {
        let mut architect_messages: Vec<Message> = messages.clone();
        let mut plan: Option<String> = None;

        let mut a_opts = loop_options.clone();
        a_opts.plan_mode = true;
        a_opts.tool_filter = Some(Arc::new(|tools: &[crate::types::ToolDefinition]| {
            read_only_tool_filter(tools)
        }));

        {
            let mut stream = run_agent_loop(architect_provider, registry.clone(), &mut architect_messages, a_opts);
            while let Some(event) = stream.next().await {
                if let AgentEvent::PlanComplete { plan: p } = &event {
                    plan = Some(p.clone());
                }
                yield event;
            }
        }

        let plan = match plan {
            Some(p) => p,
            None => {
                yield AgentEvent::Error { message: "Architect phase produced no plan".to_string() };
                return;
            }
        };

        if let Some(cb) = &architect_options.on_plan_ready {
            if !cb(plan.clone()).await {
                yield AgentEvent::Error { message: "Plan was rejected".to_string() };
                return;
            }
        }

        let mut editor_messages: Vec<Message> = messages.clone();
        editor_messages.push(Message::User(UserMessage {
            content: format!("Execute the following plan:\n\n{plan}"),
        }));

        let mut e_opts = loop_options.clone();
        e_opts.plan_mode = false;
        e_opts.tool_filter = None;

        {
            let mut stream = run_agent_loop(editor_provider, registry, &mut editor_messages, e_opts);
            while let Some(event) = stream.next().await {
                yield event;
            }
        }
        *messages = editor_messages;
    })
}
